src/precice/impl/DataContext.cpp

```diff
@@ -1,13 +1,103 @@
 #include "precice/impl/DataContext.hpp"
 #include <memory>
 #include "mesh/Data.hpp"
+#include "mesh/Mesh.hpp"
 
 namespace precice {
 namespace impl {
 
-std::string DataContext::getName() const
+DataContext::DataContext(mesh::PtrData data, mesh::PtrMesh mesh)
 {
-  return fromData->getName();
+  PRECICE_ASSERT(data);
+  _participantData = data;
+  PRECICE_ASSERT(mesh);
+  _mesh = mesh;
+}
+
+mesh::PtrData DataContext::participantData()
+{
+  PRECICE_ASSERT(_participantData);
+  return _participantData;
+}
+
+std::string DataContext::getParticipantDataName() const
+{
+  PRECICE_ASSERT(_participantData);
+  return _participantData->getName();
+}
+
+int DataContext::getParticipantDataID() const
+{
+  PRECICE_ASSERT(_participantData);
+  return _participantData->getID();
+}
+
+mesh::PtrData DataContext::fromData()
+{
+  PRECICE_ASSERT(_fromData);
+  return _fromData;
+}
+
+std::string DataContext::getFromDataName() const
+{
+  PRECICE_ASSERT(_fromData);
+  return _fromData->getName();
+}
+
+int DataContext::getFromDataID() const
+{
+  PRECICE_ASSERT(_fromData);
+  return _fromData->getID();
+}
+
+mesh::PtrData DataContext::toData()
+{
+  PRECICE_ASSERT(_toData);
+  return _toData;
+}
+
+std::string DataContext::getToDataName() const
+{
+  PRECICE_ASSERT(_toData);
+  return _toData->getName();
+}
+
+int DataContext::getToDataID() const
+{
+  PRECICE_ASSERT(_toData);
+  return _toData->getID();
+}
+
+std::string DataContext::getMeshName() const
+{
+  PRECICE_ASSERT(_mesh);
+  return _mesh->getName();
+}
+
+int DataContext::getMeshID() const
+{
+  PRECICE_ASSERT(_mesh);
+  return _mesh->getID();
+}
+
+void DataContext::setMapping(MappingContext mappingContext, mesh::PtrData fromData, mesh::PtrData toData)
+{
+  PRECICE_ASSERT(!hasMapping());
+  PRECICE_ASSERT(fromData);
+  PRECICE_ASSERT(toData);
+  _mappingContext = mappingContext;
+  _fromData       = fromData;
+  _toData         = toData;
+}
+
+bool DataContext::hasMapping() const
+{
+  return _toData && _fromData && (_toData != _fromData);
+}
```

Comment:
Wouldn't it be nicer to have a member `_hasMapping`, which is set in `setMapping` and which we could assert in `getFromData` etc?

Reply:
Currently there are two places that need to publicly access `hasMapping()`:

```
void SolverInterfaceImpl::resetWrittenData()
{
  PRECICE_TRACE();
  for (DataContext &context : _accessor->writeDataContexts()) {
    context.participantData()->toZero();
    if (context.hasMapping()) {
      PRECICE_ASSERT(context.fromData() == context.participantData());
      context.toData()->toZero();
    }
  }
}
```
see [here](https://github.com/BenjaminRodenberg/precice/blob/3ed6d071fb0417410269db2fd6cc46ce00207b16/src/precice/impl/SolverInterfaceImpl.cpp#L1603).

We could use something similar to [this code](https://github.com/BenjaminRodenberg/precice/blob/3ed6d071fb0417410269db2fd6cc46ce00207b16/src/precice/impl/SolverInterfaceImpl.cpp#L1502): `bool hasMapping = context.mappingContext().mapping.get() != nullptr;` as a workaround. But to me this does not look very explicit.

In 1ebca03f20d26cdd4a808cde0ae5ae1dc8be316c I cleaned up the rest and added some assertions as proposed. I actually replaced the code snipped from above with a call of `hasMapping()`.

Reply:
Looks good.
You could directly set
```
bool _hasMapping = false;
```
in `DataContext.hpp` instead of setting it in the constructor.

Reply:
Done in 7b2a73760dce1004d8eac10b52ca22c9db314ca6